#7363: L1 small region allocator support

nsmithtt opened the pull request:
Reserve a special region of L1 at the top for config tensors / other sideband data for ops. Usually these are tiny allocations and are long lived, so as ops are created on the fly it can introduce bad fragmentation.

Code Changes:
- Inside of struct Allocator introduce a new BankManager, l1_small_manager
- During Device::initialize_allocator a reserved carve out size is passed in, by default it could be something like 32k, but could be overridden by the user at device creation, might have to bump it up for larger models. This size is subtracted from the l1_manager’s size and feeds the l1_small_manager size. L1 managers will need to program their respective alloc_offset accordingly
- Introduce a new BufferType called L1_SMALL, tensor/buffer creation calls in ttlib need to pass down a flag to denote that they want to allocate from the reserved region
- allocator::allocate_buffer and friends just handle a new case statement for dispatching to the l1_small_manager

tt-rkim reviewed (approved):
Fixture very clean very nice!

nsmithtt commented on conftest.py:
```diff
@@ -256,15 +256,19 @@ def reset_tensix(request, silicon_arch_name):
 
 @pytest.fixture(scope="function")
-def device_init_destroy(request):
+def device_l1_small_size(request):
```
I discussed the naming with @davorchap, I originally called it `L1_RESERVED` but it was confusing because there's already `L1_UNRESERVED`.  Another name I thought of was `L1_STATIC`.  I'm not tied to the name small, we just needed a name that doesn't collide with an existing concept.  That said, this region is intended for small allocations, so as to not have them interleaved with bigger ones and causing fragmentation problems.

nsmithtt committed:
#7363: L1 small region allocator support

Reserve a special region of L1 at the top for config tensors / other
sideband data for ops. Usually these are tiny allocations and are long
lived, so as ops are created on the fly it can introduce bad
fragmentation.

Code Changes:
- Inside of struct Allocator introduce a new BankManager, l1_small_manager
- During Device::initialize_allocator a reserved carve out size is passed
  in, by default it could be something like 32k, but could be overridden
  by the user at device creation, might have to bump it up for larger
  models. This size is subtracted from the l1_manager’s size and feeds
  the l1_small_manager size. L1 managers will need to program their
  respective alloc_offset accordingly
- Introduce a new BufferType called L1_SMALL, tensor/buffer creation
  calls in ttlib need to pass down a flag to denote that they want to
  allocate from the reserved region
- allocator::allocate_buffer and friends just handle a new case statement
  for dispatching to the l1_small_manager

ntarafdar reviewed (approved):
LGTM! 

Wanted a couple clarifications though. Can we use this space for sharded tensors where individual shards are very small? I think as a short hand if we have a repeated config we do use a sharded tensor as a quick way to distribute the config tensor across all cores.

I think it makes sense at the allocater level to call them small , and this is out of the scope of this PR but do you see us maybe adding a config flag to the tensor object for this, or a new tensor object called a ConfigTensor that uses this allocater?

nsmithtt commented:
@tarafdarTT 

> Wanted a couple clarifications though. Can we use this space for sharded tensors where individual shards are very small? I think as a short hand if we have a repeated config we do use a sharded tensor as a quick way to distribute the config tensor across all cores.

Yes exactly, the 32kB is 32kB per core, meaning that technically a sharded tensor on 4 cores could be 128k total and fit in this region.

> I think it makes sense at the allocater level to call them small , and this is out of the scope of this PR but do you see us maybe adding a config flag to the tensor object for this, or a new tensor object called a ConfigTensor that uses this allocater?

Do you have any use cases in mind where this could be useful?

At first glance I don't see a need to specialize config tensors, theoretically there could be a config tensor that's too large and doesn't make sense to allocate from `L1_SMALL`. `L1_SMALL` is just another memory space that tensors can live, just like `L1` or `DRAM` or `SYSTEM` and I think the tensor should remain agnostic to its underlying storage.

rtawfik01 commented on tt_metal/third_party/tt_llk_wormhole_b0:
Hey @nsmithtt , you need to run 
```
git submodule update --init --recursive
```
The PR is accidentally reverting the submodule pointer

nsmithtt replied:
Good catch, thank you! Reverting that now

nsmithtt replied:
@rtawfik01, reverted
